logger.go

```diff
@@ -5,155 +5,260 @@ import (
 	"os"
 	"runtime"
 	"strings"
+	"sync"
+	"testing"
 	"time"
 )
 
-type logWrappedT struct {
-	T
+type loggerT[ET T] struct {
+	runTHelper    // Embeds T and provides Fail/Parallel
+	orig       ET // Keep reference to original for Run method
+	logger     func(string)
+	skipFrames int // Additional skip frames for nested wrappers
+}
+
+// replaceLoggerT directly implements T interface to avoid double loggerT wrapping
+type replaceLoggerT[ET T] struct {
+	runTHelper
+	orig   ET
 	logger func(string)
 }
 
-// ReplaceLogger creates a T that is wrapped so that the logger is
-// overridden with the provided function.
-func ReplaceLogger(t T, logger func(string)) T {
-	return logWrappedT{
-		T:      t,
-		logger: logger,
+// ReplaceLogger creates a logger wrapper that overrides the logging function. When layered
+// on top of BufferedLogger, it assumes that only one extra stack frame is added. If that's
+// not the case, cast and adjust:
+//
+//	if asf, ok := t.(interface{ AdjustSkipFrames(int) }); ok {
+//		asf.AdjustSkipFrames(2)
+//	}
+//
+// This adjustment should be done before using the the returned T
+func ReplaceLogger[ET T](t ET, logger func(string)) T {
+	// If the underlying logger supports AdjustSkipFrames, adjust it to account for
+	// the extra call frames: replaceLoggerT.Log -> custom logger function -> underlying logger call
+	if adjuster, ok := any(t).(interface{ AdjustSkipFrames(int) }); ok {
+		adjuster.AdjustSkipFrames(2) // +2 for replaceLoggerT.Log and the custom logger function
+	}
+
+	wrapped := &replaceLoggerT[ET]{
+		runTHelper: runTHelper{T: t},
+		orig:       t,
+		logger:     logger,
 	}
+	return any(wrapped).(T)
```

Comment:
```suggestion
	return T(wrapped)
```

Untested but I'd expect this to work.  If there's a reason for `any(wrapped).(T)` I'm curious to know what it is.